python/dgl/sparse.py

```diff
@@ -190,7 +190,49 @@ def _gspmm(gidx, op, reduce_op, u, e):
 
 
 def _gspmm_hetero(gidx, op, reduce_op, u_len, u_and_e_tuple):
-    r""" Generalized Sparse Matrix Multiplication interface.
+    r""" Generalized Sparse Matrix Multiplication interface on heterogeneous graphs.
+    It handles multiple node and edge types of the graph. For each edge type, it takes
+    the result of :attr:`op` on source node feature and edge feature, and leads to a
+    message on edge. Then it aggregates the message by :attr:`reduce_op` on the destination
+    nodes of the etype.
+
+    .. math::
+        x_v = \psi_{(u, v, e)\in \mathcal{G}}(\rho(x_u, x_e))
+
+    where :math:`x_v` is the returned feature on destination nodes, and :math`x_u`,
+    :math:`x_e` refers to :attr:`u`, :attr:`e` respectively. :math:`\rho` means binary
+    operator :attr:`op` and :math:`\psi` means reduce operator :attr:`reduce_op`,
+    :math:`\mathcal{G}` is the graph we apply gspmm on: :attr:`g`.
+
+    Note that this function does not handle gradients.
+
+    Parameters
+    ----------
+    gidx : HeteroGraphIndex
+        The input graph index.
+    op : str
+        The binary op's name, could be ``add``, ``sub``, ``mul``, ``div``, ``copy_lhs``,
+        ``copy_rhs``.
+    reduce_op : str
+        Reduce operator, could be ``sum``, ``max``, ``min``.
+    u_len : int
+        The number of tensors in ``u`` (source node features)
+    u_and_e_tuple : Tuple of tensors
+        Tuple of source nodes' features and edges' features.
+        The feature on source nodes could be None if op is ``copy_rhs``.
+        The feature on edges, could also be None if op is ``copy_lhs``.
```

Comment:
```
The first few elements are source nodes' features, one for each node type.  If
``op`` is ``copy_rhs``, all of them can be None.
The rest of them are edges' features, one for each edge type.  If ``op`` is ``copy_lhs``,
all of them can be None.
```
Also, say if `op` is `copy_rhs`, we set *all* of the source node features to be None, correct?

Reply:
Clarified the comments.